icu4c/source/common/unicode/brkiter.h

```diff
@@ -93,6 +95,34 @@ U_NAMESPACE_BEGIN
  * are available at http://www.unicode.org/reports/tr14/ and
  * http://www.unicode.org/reports/tr29/.
  * <p>
+ * A boundary logically occurs between two adjacent code points. The boundary
+ * analysis functions in this file return the index of the code unit directly
+ * after a boundary - so a value of 0 means "the beginning of the string" and a
+ * value of u_strlen(string) means "the end of the string".
+ * <p>
+ * A span of text between two adjacent boundaries is called a "segment."
+ * The boundary analysis functions in this file behave correctly regardless of
+ * whether the current position is set to a boundary location or into the
+ * middle of a segment. When the current position points to a boundary, the two
+ * segments that straddle that boundary are called the "previous segment" and
+ * the "current segment."
+ * <p>
+ * If the current position is set to the middle of a surrogate pair, the
```

Comment:
There's an interesting difference between ICU4C and ICU4J here. See [ICU-13361](https://unicode-org.atlassian.net/browse/ICU-13361). The issue is with `preceding(index)` when index points to a trail surrogate, and there is a boundary preceding the code point. We never reached a consensus on which behavior to choose, and I'm concerned changing either one (C or J) will break somebody. My inclination would be to keep ICU4C's behavior, and change ICU4J to match.